Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -7594,6 +7602,12 @@ bool WebPageProxy::hasAllowedToRunInTheBackgroundActivity() const
     return internals().pageAllowedToRunInTheBackgroundActivityDueToTitleChanges || internals().pageAllowedToRunInTheBackgroundActivityDueToNotifications;
 }
 
+void WebPageProxy::dropAllowedToRunInTheBackgroundActivities()
+{
+    internals().pageAllowedToRunInTheBackgroundActivityDueToTitleChanges = nullptr;
+    internals().pageAllowedToRunInTheBackgroundActivityDueToNotifications = nullptr;
```

Comment:
Does this mean that if a website with notifications needs to be suspended aggressively we will no longer get notifications?

I think the trade-off between no notifications and crashing might be correct but I just want to understand if that's what's going on here.

Reply:
The existing heuristic basically makes it so that pages that use local notifications (such as some messaging apps) effectively never suspend. After this patch, if that page crosses the BG mem threshold, then it won't continue running indefinitely.

If the page is using Web Push, then that's not affected, as the service worker for the page handles the push event even if the page itself is suspended.